src/state/ledger/user_commands.rs

```diff
@@ -0,0 +1,117 @@
+use mina_serialization_types::staged_ledger_diff::{
+    self, SignedCommandPayloadBody, StakeDelegation, TransactionStatus,
+};
+
+use crate::block::precomputed::PrecomputedBlock;
+
+use super::public_key::PublicKey;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Copy, Clone, Hash)]
+pub enum UserCommandType {
+    Payment,
+    Delegation,
+}
+
+pub struct BalanceUpdate {
+    pub public_key: PublicKey,
+    pub balance: u64,
+}
+
+pub struct UserCommand {
+    pub source_nonce: i32,
+    pub command_type: UserCommandType,
+    pub fee_payer: BalanceUpdate,
+    pub source: BalanceUpdate,
+    pub receiver: BalanceUpdate,
+}
+
+impl UserCommand {
+    pub fn from_precomputed(precomputed_block: &PrecomputedBlock) -> Vec<Self> {
+        precomputed_block
+            .staged_ledger_diff
+            .diff
+            .t
+            .0
+            .t
+            .t
+            .commands
+            .iter()
+            .map(|command| command.t.clone())
+            .flat_map(|command| {
+                let source_nonce = match &command.data.t.t {
```

Comment:
All these `.t`s are gross! Maybe we can isolate all these accessors nicely into their own functions?

Reply:
I mostly took care of this